yggdrasilffi/src/lib.rs

```diff
@@ -56,257 +63,241 @@ impl From<serde_json::Error> for FFIError {
     }
 }
 
-#[no_mangle]
-pub extern "C" fn engine_new() -> *mut c_void {
-    let state = EngineState::default();
-    Box::into_raw(Box::new(state)) as *mut c_void
+fn get_engine<'a>(engine_ptr: *mut c_void) -> Result<&'a mut EngineState, FFIError> {
```

Comment:
I would rather we mark the function as unsafe itself, rather than the block of code that does the unsafe stuff. 

Putting the unsafe on the function level tells other programmers that "you need to know what you're doing here, if you pass the wrong thing or do the wrong thing with the return type, bad things may happen". In this case, the wrong thing can be passing a valid pointer to an invalid reference. 

Putting the unsafe inside the function without marking the function as unsafe tells other programmers "there might be some unsafe stuff in here, but don't worry, we did checking before and there's no way you, as the caller, can cause the bad things to happen"

Reply:
(but to clear, this is a really nice cleanup)

Reply:
That's fair. Addressed in https://github.com/Unleash/yggdrasil/pull/47/commits/c7723e0b49e4a941e3dab9a39dcd39b3c3c64ded

My reasoning for this change was that it felt better to me to only specify `unsafe` when needed. Marking a method as `unsafe` means you can technically do anything unsafe inside that method and the compiler won't scream at you, which feels wrong. At that point, do we still mark the unsafe code inside the method as unsafe or not? Whatever we decide to do, the compiler won't care, which again, feels wrong to me.

However I understand that from an interface perspective it may make sense to mark the methods as unsafe.

Reply:
It does make a lot of sense from the interface perspective, unfortunately. Our methods here are definitely not safe and we can't make them safe, so we should be polite enough to tell other people's documentation/compilers that that's the case. Otherwise we run the risk of letting others call these methods blindly without understanding the consequences (or even that there are consequences).

What I was originally going for is something like

```fn unsafe some_public_method() { 
  let results = unsafe {
  // some stuff that may set fire to the universe
  }
  handle_stuff(results) //some safe method, where we can write polite code
}
```

But I don't know if that strategy worked out. It got a bit too messy.

I think what you have at this point is a pretty good balance. Kinda want to just merge it